File: Sources/UsageEngine/CachePricing.swift
import Foundation

/// Approximate cache-savings estimator (Claude-only).
/// Rates are representative LiteLLM USD-per-million-token values, hardcoded
/// because this figure is an illustrative "you saved ≈$Y" number, not billing.
public enum CachePricing {
    /// (inputPerMTok, cacheReadPerMTok) by model-name prefix. First match wins.
    private static let rateTable: [(prefix: String, input: Double, cacheRead: Double)] = [
        ("claude-opus",   15.0, 1.50),
        ("claude-sonnet",  3.0, 0.30),
        ("claude-haiku",   1.0, 0.10),
    ]

    public static func estimatedSavings(cacheReadTokens: Int, model: String) -> Double {
        guard ToolClassifier.tool(forModel: model) == .claude, cacheReadTokens > 0 else { return 0 }
        let name = model.lowercased()
        guard let rate = rateTable.first(where: { name.hasPrefix($0.prefix) }) else { return 0 }
        let perToken = (rate.input - rate.cacheRead) / 1_000_000.0
        return Double(cacheReadTokens) * perToken
    }
}

File: windows/internal/engine/cachepricing.go
package engine

import "strings"

// cacheRate is (input, cacheRead) USD per million tokens for a model-name prefix.
type cacheRate struct {
	prefix    string
	input     float64
	cacheRead float64
}

// cacheRateTable holds representative LiteLLM USD-per-million-token values,
// hardcoded because cache savings is an illustrative "you saved ~$Y" figure, not
// billing. First matching prefix wins.
var cacheRateTable = []cacheRate{
	{"claude-opus", 15.0, 1.50},
	{"claude-sonnet", 3.0, 0.30},
	{"claude-haiku", 1.0, 0.10},
}

// CacheSavings estimates the dollars saved by prompt caching across the given
// model breakdowns. Claude-only and approximate.
func CacheSavings(models []ModelBreakdown) float64 {
	total := 0.0
	for _, m := range models {
		total += cacheSavingsForModel(m.CacheReadTokens, m.ModelName)
	}
	return total
}

// cacheSavingsForModel is the per-model estimate: cache-read tokens billed at the
// cache-read rate instead of the full input rate.
func cacheSavingsForModel(cacheReadTokens int64, model string) float64 {
	if ClassifyModel(model) != ToolClaude || cacheReadTokens <= 0 {
		return 0
	}
	name := strings.ToLower(model)
	for _, r := range cacheRateTable {
		if strings.HasPrefix(name, r.prefix) {
			return float64(cacheReadTokens) * (r.input - r.cacheRead) / 1_000_000.0
		}
	}
	return 0
}
